String solver: stop relying on array size

allredj opened the pull request:
The string's backing array size (from its type) initially informs about the length of the string. However during refinement, the length can vary and if the length is tied to the size of the array, we need to rewrite the type as a conditional, which should be avoided (both the rewriting and the conditional).

This PR refactors string refinement so that during constraint generation, we refer to the size of the string solely through the array pool.
The size of a string thus always refers to a symbol, which may or may not be constrained.

The only time we get the array size is when inserting an array_string_exprt in the array pool (i.e. at the creation of the array_string_exprt).

There shouldn't be any visible change in JBMC, but that should fix some errors that have been happening with test-generation. I will keep looking to see if I can reproduce on JBMC.

I will also check that performance is not affected.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -32,6 +32,18 @@ exprt array_poolt::get_length(const array_string_exprt &s) const
   return s.length();
 }
 
+/// As opposed to get_length(), do not create a new symbol if the length
```
:pick: `\ref`

romainbrenguier replied:
this is part of the public interface so should be documented in the header file :grapes:

allredj replied:
`\ref` is unnecessary. Doxygen understands where to link.

allredj replied:
Doc moved to header file

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -21,15 +21,14 @@ operator()(const irep_idt &prefix, const typet &type)
 
-exprt array_poolt::get_length(const array_string_exprt &s) const
+/// Get the length of an array_string_exprt from the array_pool. If the length
+/// does not yet exist, create a new symbol and add it to the pool.
```
:grapes:

allredj replied:
Doc moved to header file

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -21,15 +21,14 @@ operator()(const irep_idt &prefix, const typet &type)
 }
 
-exprt array_poolt::get_length(const array_string_exprt &s) const
+/// Get the length of an array_string_exprt from the array_pool. If the length
```
:pick: typo in commit message: "accesssing"

allredj replied:
Fixed

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -46,8 +46,7 @@ array_poolt::get_length_if_exists(const array_string_exprt &s) const
 {
-  symbol_exprt length = fresh_symbol("string_length", index_type);
-  array_typet array_type(char_type, length);
+  array_typet array_type(char_type, infinity_exprt(index_type));
```
:pick: brace initialization

allredj replied:
Done

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -23,8 +23,21 @@ operator()(const irep_idt &prefix, const typet &type)
 {
+  if(const auto &if_expr = expr_try_dynamic_cast<if_exprt>((exprt)s))
+  {
+    return if_exprt(
```
:pick: brace initialization

allredj replied:
Done

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -115,6 +115,54 @@ array_string_exprt array_poolt::make_char_array_for_char_pointer(
+        length_of_array.emplace(array_expr, size_from_type);
+      INVARIANT(
+        emplace_result.second,
+        "attempt_assign_length should only be called when no entry for the"
```
`attempt_assign_length_from_type`

allredj replied:
Fixed

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -65,6 +65,10 @@ array_poolt::fresh_string(const typet &index_type, const typet &char_type)
     address_of_exprt(index_exprt(str, from_integer(0, index_type))), str);
+
+  // add length to length_of_array map
+  get_length(str);
```
:pick: maybe worth defining a void function to avoid having to write that comment twice

allredj replied:
Maybe in a later PR.

romainbrenguier commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -184,9 +184,12 @@ void array_poolt::insert(
-const std::set<array_string_exprt> &array_poolt::created_strings() const
+/// Return a map mapping all array_string_exprt of the array_pool to their
+/// length.
+const std::unordered_map<array_string_exprt, exprt, irep_hash> &
```
:bulb: exposing the whole map here seems unnecessary since we are only using the first component. An alternative would be to define a `for_each_string` function with takes a function as argument and apply it to each `array_string_exprt`.

allredj replied:
I'll leave it like this for now so that the PR doesn't blow up.

romainbrenguier reviewed (commented):
Reviewed up to "Get rid of created_strings_value". No blocking request so far.

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -1834,9 +1834,12 @@ exprt string_refinementt::get(const exprt &expr) const
+        generator.array_pool.get_length_if_exists(arr))
     {
-      const exprt length = super_get(arr.length());
+      exprt length = super_get(length_from_pool.value());
```
no need to remove the `const`

allredj replied:
Re-introduced

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -105,7 +105,8 @@ static optionalt<exprt> get_array(
   const namespacet &ns,
   messaget::mstreamt &stream,
-  const array_string_exprt &arr);
+  const array_string_exprt &arr,
```
:pick: commit message doesn't say why

allredj replied:
Added description

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -965,7 +965,11 @@ static optionalt<exprt> get_array(
 {
-  const exprt &size = arr.length();
+  const auto &size_from_pool = array_pool.get_length_if_exists(arr);
+  exprt size = size_from_pool.has_value()
```
no need to remove the `const`

allredj replied:
const re-introduced

romainbrenguier commented on src/solvers/strings/string_constraint_generator_format.cpp:
```diff
@@ -493,8 +498,9 @@ std::pair<exprt, string_constraintst> add_axioms_for_format(
         const array_string_exprt string_arg = get_string_expr(array_pool, arg);
-        get_arg = [string_arg](const irep_idt &id) {
-          return format_arg_from_string(string_arg, id);
+        /// REMOVE get_arg below
```
forgotten comment?

allredj replied:
Removed

smowton commented:
Will look at this tomorrow, but do you intend it to remove the illegal `x ? y : z` expressions where x has type `char[x ? sizeof(y) : sizeof(z)]`?

smowton commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -32,6 +32,18 @@ exprt array_poolt::get_length(const array_string_exprt &s) const
 }
 
+/// As opposed to get_length(), do not create a new symbol if the length
+/// of the array_string_exprt does not exist, but instead return an empty
```
Perhaps clarify what you mean by a length not existing? Is that any non-constant length, or the length of a string we haven't fully initialised yet, or...?

allredj replied:
Improved the documentation. This now states: 
```
  /// As opposed to get_length(), do not create a new symbol if the length
  /// of the array_string_exprt does not have one in the array_pool, but instead
  /// return an empty optional.
```

smowton commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -21,15 +21,14 @@ operator()(const irep_idt &prefix, const typet &type)
-      return it->second;
-  }
-  return s.length();
+  auto emplace_result =
```
Given this allocates a symbol even if `s` already has a length, I'd suggest inserting a dummy and then replacing it with a `fresh_symbol` if appropriate

allredj replied:
Done in fixup commit

smowton commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -115,6 +115,54 @@ array_string_exprt array_poolt::make_char_array_for_char_pointer(
+  std::unordered_map<array_string_exprt, exprt, irep_hash> &length_of_array,
+  symbol_generatort &symbol_generator)
+{
+  if(const auto &if_expr = expr_try_dynamic_cast<if_exprt>((exprt)array_expr))
```
Why the explicit cast (and to a non-reference-type?)

allredj replied:
You mean the `(exprt)`? Curiously `exprt_try_dynamic_cast` doesn't work directly on `array_string_exprt`.

smowton replied:
Oh wait, this is obvious -- an `array_string_exprt` is certainly not an `if_exprt`, this can't ever be run.

smowton replied:
There's one other case with the same cast and the same unreachable code. This suggests writing tests to exercise this code!

allredj replied:
I don't think it's unreachable. `array_string_exprt`s can have `ID_if` AFAIK. However you make a good point with the test coverage, as we don't seem to have anything to cover these lines so for now I might just put an assertion.

allredj replied:
So in this place i replaced the if with an assertion because I can't find any test to cover this part. However in the other place where I use the condition, I left it in because it's often the case that the `array_string_exprt` is an if. Therefore that code is well covered by regression tests.

smowton replied:
Aha, my mistake, I thought `array_string_exprt` constrained the expression to have a particular id, and therefore not be an `if_exprt`, but actually it just constrains the type of a general expression. The implementation of the expr-casting stuff, though, assumes this sort of trait class doesn't exist, and so casts are only possible to children of the current type. That ought to be fixed, but is obviously out of scope for this PR, so approving.

smowton commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -115,6 +115,54 @@ array_string_exprt array_poolt::make_char_array_for_char_pointer(
+    }
+    else
+    {
+      length_of_array.emplace(
```
Should check this one is new, too? Therefore, deduplicate the `emplace` calls and the invariant below this if block?

allredj replied:
Done in fixup commit

smowton commented on src/solvers/strings/array_pool.cpp:
```diff
@@ -65,6 +65,10 @@ array_poolt::fresh_string(const typet &index_type, const typet &char_type)
     address_of_exprt(index_exprt(str, from_integer(0, index_type))), str);
+
+  // add length to length_of_array map
+  get_length(str);
```
I wouldn't normally expect a `get_` function to have a side-effect like this, suggest renaming

allredj replied:
Would you be happy with `get_or_create_length`?

smowton replied:
Sounds good to me

allredj replied:
Renamed

smowton commented on src/solvers/strings/array_pool.h:
```diff
@@ -76,6 +76,8 @@ class array_poolt final
 
 private:
   /// Associate arrays to char pointers
+  /// Invariant: Each array_string_exprt in this map has a corresponding entry
```
Where is this enforced?

allredj replied:
Improved documentation. Now states:
```
  /// Associate arrays to char pointers
  /// Invariant: Each array_string_exprt in this map has a corresponding entry
  ///            in length_of_array.
  ///            This is enforced whenever we add an element to
  ///            `arrays_of_pointers`, i.e. fresh_string()
  ///            and make_char_array_for_char_pointer().
```

allredj commented:
@smowton About your general comment. Yes, removing the conditional types is the idea of that PR.

tautschnig reviewed (approved):
Great to see those conditionals in types go away! I'd just ask for the commit history to be cleaned up with all those fixup commits. Doing so will then also re-trigger the codebuild task.

allredj commented:
@tautschnig Thanks. Of course I'll merge the fixup commits before merge. I have some failing tests on TG, so I need to fix that first anyway.

allredj commented:
@romainbrenguier can you please review the 7 last commits which refactor `get_array` and solve the issue with the dangling string literals in the trace (in return values). I 'm still looking for a regression test for that.

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -1003,7 +1003,7 @@ static optionalt<exprt> get_array(
     stream << "(sr::get_array) this can also happen on invalid object access"
            << messaget::eom;
-    return nil_exprt();
+    return {};
```
This can potentially change the behaviour in the test `jbmc/regression/jbmc-strings/long_string/test_abort.desc` because here MAX_CONCRETE_STRING_SIZE was exceeded (on purpose) and the value of the array replaced by nil in the trace. Can you check what happens now in the trace?

allredj replied:
Indeed that broke that test. I reverted that change and consequently changed the refactoring a bit. I moved the max length check back into `get_array`.

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -949,16 +949,18 @@ void string_refinementt::add_lemma(
-/// \return an optional array expression or array_of_exprt
-static optionalt<exprt> get_array(
+/// \return an optional expression representing the size of the array that can
+///         be cast to type_t
```
what's `type_t`?

allredj replied:
I meant `size_t`. Fixed

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -970,42 +972,72 @@ static optionalt<exprt> get_array(
+
+  size_t n = numeric_cast<std::size_t>(size.value()).value();
+
+  exprt arr_val = simplify_expr(adjust_if_recursive(super_get(arr), ns), ns);
```
if the type problems are fixed by the previous commits, the calls to `adjust_if_recursive` should no longer be required

allredj replied:
I don't know. I'll leave it for now.

tautschnig replied:
We should then have a follow-up PR that removes `adjust_if_recursive` I'd say.

allredj replied:
@tautschnig https://github.com/diffblue/cbmc/pull/4653

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -970,42 +972,72 @@ static optionalt<exprt> get_array(
+    return {};
+  }
+
+  size_t n = numeric_cast<std::size_t>(size.value()).value();
```
const?

allredj replied:
done

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -970,42 +972,72 @@ static optionalt<exprt> get_array(
+  const array_string_exprt &arr,
+  const array_poolt &array_pool)
+{
+  auto size = get_valid_array_size(super_get, ns, stream, arr, array_pool);
```
const

allredj replied:
done

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -970,42 +972,72 @@ static optionalt<exprt> get_array(
+
+  size_t n = numeric_cast<std::size_t>(size.value()).value();
+
+  exprt arr_val = simplify_expr(adjust_if_recursive(super_get(arr), ns), ns);
```
const

allredj replied:
done

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -982,13 +982,13 @@ static optionalt<exprt> get_valid_array_size(
 
-  auto n_opt = numeric_cast<std::size_t>(size_val);
-  if(!n_opt)
+  auto num_size_opt = numeric_cast<std::size_t>(size_val);
```
`num_size_opt` is also not a great name (what does "num" stand for?). `size_opt` could be good enough?

allredj replied:
Got rid of that change

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -988,12 +988,12 @@ static optionalt<exprt> get_valid_array_size(
     return {};
   }
-  std::size_t n = *num_size_opt;
+  std::size_t num_size = *num_size_opt;
```
:pick: I would rather name this one `size` and the previous size `size_expr`

allredj replied:
Got rid of that change

Commit:
Add array_pool as a field of string_builtin_functionst

This is necessary in preparation for the upcoming change where all
accesses to the length of strings must be done through the array pool.

Commit:
get_length_if_exists() function

Variant of get_length that does not create a new symbol.
This will become useful for the first stages of string refinement, where
we want to be sure that the length of array_string_exprt actually
exists.

Commit:
Don't set the array type's length

The length of the string's backing array should not be read during
constraint generation.
Therefore it isn't necessary to set it.

Commit:
array_poolt.get_length() returns a new symbol by default

In the event the length of the queried array_string_poolt does not yet
exist, create a new symbol for the length and store it in the pool.

This avoids accessing the length of the backing array directly, which
is unreliable.

Commit:
fresh_string: do not create a length symbol

The length of the backing array of strings should not be read, so no
need to create a length symbol for the type when creating a new string.

Commit:
Open length_of_array to non-symbol lengths

Lengths of array_string_exprt can be any exprt, not just symbol_exprt.
In particular, we want to be able to give the length coming from the
backing array, which is not necessarily a symbol.

Commit:
Make array_poolt.get_length handle if_exprts

Commit:
Add missing entry to arrays_of_pointers map

All created strings should have an entry in arrays_of_pointers.

Commit:
Add attempt_assign_length_from_type

Function to add the length of the string backing array to the
length_of_array map in the array_pool. If the size does not exist,
create a symbol.

Function will be used to assign an initial length to the
array_string_exprt, using the type information.

After that initial assignment, there will be no further reference to the
length of the underlying array.

Commit:
fixup! Add attempt_assign_length_from_type

Remove handling of if_case, as I cannot find a test case that exercises
it.

Commit:
Use attempt_assign_length_from_type() in array_poolt.insert()

Commit:
Make arg of array_pool.insert() const

Because we can.

Commit:
Document and enforce invariant in array_pool

This invariant ensures that we get all the strings of the array_pool
when iterating over length_of_array, allowing to get rid of the
created_strings_value map.

Commit:
array_pool.created_strings() now uses length_of_array

Using the newly adding invariants, created_strings() now returns the
length_of_array map, which contains all array_string_exprts of the
array_pool.

This map is more complete than created_strings_value.

Commit:
Get rid of created_strings_value

No longer used.

array_poolt.find() can be made to return the array_string_exprt directly

Commit:
string_refinement::get(): get length if exists

In string_refinement::get(), we now get the length of the string from
the array_pool instead of the array type.

We use get_length_if_exists() instead of get_length(), because we want to
ignore if the length does not exist, instead of creating a new
symbol.

Commit:
Pass array_pool arg to get_array

String length will be queried from array_pool in future commit.

Commit:
Get length from array_pool in get_array()

Avoid reading from array size.

Commit:
Propagate array_pool to constraint generation

Argument will be required to query string lengths.

Commit:
Read string length from array pool

Replace calls to str.length() by calls to array_pool.get_length(str).

Commit:
Unit test: get constant length directly from array

We require a constant value in this case and querying the array pool
gives us a symbol.

Commit:
Remove unused array_string_exprt.length()

Length is now always read from array_pool

Commit:
Refactor format support to not use get_arg function argument

Simplifies the interface by passing that string argument directly rather
than the function that retrieves it.

Commit:
Format: Remove unnecessary conversion

Commit:
Rename get_length to get_or_create_length

Commit:
get_array: remove useless variables

Commit:
Renmove ref to size in warning message

size exprt is not meaningful in all cases.
This is necessary for upcoming refactoring.

Commit:
Refactor get_array: extract get_valid_array_size

Commit:
Get size from array if not available in model

If the size of the array is not provided by the model, get the size
directly from the array.

Note that the non-constant case now returns an empty optional rather
than an unknown exprt.

Commit:
Test for concretized string literal in trace

Checks that the string literal is present in the trace even if it is not
part of the decision. As a side-effect, this checks that the length is
properly assigned to the string, even if it cannot be retrieved from the
solver.

allredj commented:
Fixed outstanding issues and addressed comments. @romainbrenguier could you re-review the 5 last commits please?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6da6e01b7d8f039f789e1411b1eb064e8cf3aed5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111727112

romainbrenguier commented on src/solvers/strings/string_refinement.cpp:
```diff
@@ -993,8 +993,8 @@ static optionalt<exprt> get_array(
   {
-    stream << "(sr::get_array) long string (size " << format(size) << " = " << n
-           << ") " << format(arr) << messaget::eom;
+    stream << "(sr::get_array) long string (size = " << n << ") " << format(arr)
```
:pick: typo in commit message: "Renmove"

romainbrenguier reviewed (approved):
Looks good

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 09588ad748d4528da901958b49fdbf7618f298d5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112432177
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 27e27c0ee3860c446afcff7b63843ce99b827b1c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113065638

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: cd61e81953ec41d7235f767ac9e1906fac8a137c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113725722
